File: java/builder/ProductService.java
package builder;

import java.io.IOException;
import java.net.URI;
import java.net.URISyntaxException;
import java.net.http.HttpRequest;
import java.net.http.HttpResponse;
import java.time.Duration;

import static java.time.temporal.ChronoUnit.SECONDS;

public class ProductService {

    public HttpResponse<String> getProducts() throws URISyntaxException, IOException, InterruptedException {
        HttpRequest request = HttpRequest.newBuilder()
                .uri(new URI("https://dummyjson.com/products/1"))
                .headers("key1", "value1", "key2", "value2")
                .timeout(Duration.of(10, SECONDS))
                .GET()
                .build();
        java.net.http.HttpClient client = java.net.http.HttpClient.newBuilder()
                .build();
        return client.send(request, HttpResponse.BodyHandlers.ofString());
    }
}

File: C#/Builder/ProductService.cs
using Microsoft.Extensions.DependencyInjection;

namespace Builder;

public sealed class ProductService
{
    /// <summary>
    /// More commonly used in ASP.NET applications
    /// </summary>
    /// <returns></returns>
    private IHttpClientBuilder CreateBuilder()
    {
        ServiceCollection services = new();

        // Can Chain other calls for further configuration
        IHttpClientBuilder x = services.AddHttpClient<ProductService>("product")
            .ConfigureHttpClient(client =>
            {
                client.BaseAddress = new Uri("https://dummyjson.com/");
                client.DefaultRequestHeaders.Add("Accept", "application/json");
                client.Timeout = TimeSpan.FromSeconds(10);
            });

        return x;
    }

    public string GetProducts()
    {
        IHttpClientBuilder builder = CreateBuilder();

        IHttpClientFactory clientFactory = builder.Services.BuildServiceProvider()
            .GetRequiredService<IHttpClientFactory>();
            
        HttpClient client = clientFactory.CreateClient("product");
        
        // Usage of Result is not recommended in production code
        // This is just for demonstration purposes
        // use async/await instead
        var response = client.GetAsync("products/1").Result;
        var products = response.Content.ReadAsStringAsync().Result;
        return products;
    }
}
